interface/main/main_screen.php

```diff
@@ -90,87 +90,136 @@ function generate_html_end()
-            // We have to use the matching registered key.
-            $strhandle = json_encode($registration->keyHandle);
-            if (isset($regs[$strhandle])) {
+    if ($registrations["method"] == "U2F") {
+        // There is at least one U2F key registered so we have to request or verify key data.
+        // https is required, and with a proxy the server might not see it.
+        $scheme = "https://"; // isset($_SERVER['HTTPS']) ? "https://" : "http://";
+        $appId = $scheme . $_SERVER['HTTP_HOST'];
+        $u2f = new u2flib_server\U2F($appId);
+        $userid = $_SESSION['authId'];
+        if ($form_response) {
+            // We have key data, check if it matches what was registered.
+            $tmprow = sqlQuery("SELECT login_work_area FROM users_secure WHERE id = ?", array($userid));
+            try {
+                $registration = $u2f->doAuthenticate(
+                    json_decode($tmprow['login_work_area']), // these are the original challenge requests
+                    $registrations,
+                    json_decode($_POST['form_response'])
+                );
+                // Stored registration data needs to be updated because the usage count has changed.
+                // We have to use the matching registered key.
+                $strhandle = json_encode($registration->keyHandle);
+                if (isset($regs[$strhandle])) {
+                    sqlStatement(
+                        "UPDATE login_mfa_registrations SET `var1` = ? WHERE " .
+                        "`user_id` = ? AND `method` = 'U2F' AND `name` = ?",
+                        array(json_encode($registration), $userid, $regs[$strhandle])
+                    );
+                } else {
+                    error_log("Unexpected keyHandle returned from doAuthenticate(): '$strhandle'");
+                }
+                // Keep track of when challenges were last answered correctly.
                 sqlStatement(
-                    "UPDATE login_mfa_registrations SET `var1` = ? WHERE " .
-                    "`user_id` = ? AND `method` = 'U2F' AND `name` = ?",
-                    array(json_encode($registration), $userid, $regs[$strhandle])
+                    "UPDATE users_secure SET last_challenge_response = NOW() WHERE id = ?",
+                    array($_SESSION['authId'])
                 );
-            } else {
-                error_log("Unexpected keyHandle returned from doAuthenticate(): '$strhandle'");
+            } catch (u2flib_server\Error $e) {
+                // Authentication failed so we will build the U2F form again.
+                $form_response = '';
+                $errormsg = xl('Authentication error') . ": " . $e->getMessage();
             }
-            // Keep track of when challenges were last answered correctly.
+        }
+        if (!$form_response) {
+            // There is no key data yet or authentication failed, so we need to solicit it.
```

Comment:
If there are multiple MFA methods, you will need to solicit the method that they want to use. Looks like some code will need to be rewritten.